### PR TITLE
122 fix path expr annotation

### LINKED ISSUE #122: Fix the PathExpr annotation
# Issue

## Description

The PathExpr is at the moment only returning a sequence of nodes, it should return the type of the last node.

## Sub-tasks

Atomized tasks that cannot be broken down further.

- [x] Anotate the pathExpr correctly

### PR DESCRIPTION
**Pull request by maanpalle**
# Pull Request

## Description

Fixed the pathExpression annotation

Closes #122 

## Changes

1. changed the pathExpressionAnnotation to work correctly
2. Fixed the tests

## Deletes Source Branch?

Yes:    Work is done, the branch can be safely deleted.

## How Many Approvals?

* 2 Approvals (Normal features and bugs)

### CODE REVIEW

File: test/specs/annotation.tests.ts [hunk 1]
```diff
@@ -78,7 +78,7 @@ describe('Annotate unary lookup', () => {
 describe('Path expression test', () => {
 	it('Path expression test', () => {
 		// The same query also triggers the path expression
-		assertValueType('map{"num":1}[?num]', ValueType.ITEM, undefined);
+		assertValueType('map{"num":1}[?num]', ValueType.MAP, undefined);
```
Comment (FBroy) [inline test/specs/annotation.tests.ts]: We want more tests to test if the behavior you're trying to achieve is correct!

File: src/typeInference/annotatePathExpr.ts [hunk 2]
```diff
@@ -9,11 +9,29 @@ import astHelper, { IAST } from '../parsing/astHelper';
...
+		}
+		if (astHelper.followPath(lastChild, ['xpathAxis'])) {
+			seqType = {
+				type: ValueType.NODE,
+				mult: SequenceMultiplicity.ZERO_OR_MORE,
+			};
+		}
+	}
+	if (!seqType) {
+		seqType = {
```
Comment (FBroy) [inline src/typeInference/annotatePathExpr.ts]: We do not want to insert item()* into the AST. Return the seqType in this if statement.

**Review by FBroy — CHANGES_REQUESTED** [review]

File: test/specs/annotation.tests.ts (hunk 1 above)
Comment (maanpalle) [reply test/specs/annotation.tests.ts]: Most of this code is tested via other tests but I can add some specific ones

**Review by HectorPeeters — APPROVED** [review]
After you implement franks changes, it looks good to go!

File: src/typeInference/annotatePathExpr.ts [hunk 3]
```diff
@@ -9,11 +9,29 @@ import astHelper, { IAST } from '../parsing/astHelper';
...
+			type: ValueType.ITEM,
+			mult: SequenceMultiplicity.ZERO_OR_MORE,
+		};
```
Comment (yuanzexiong) [inline src/typeInference/annotatePathExpr.ts]: Is this the new treatment for the cases that we cannot infer the type? Item()* in another shell?

**Review by yuanzexiong — APPROVED** [review]

File: src/typeInference/annotatePathExpr.ts (hunk 3 above)
Comment (FBroy) [reply src/typeInference/annotatePathExpr.ts]: This represents item()*

File: src/typeInference/annotatePathExpr.ts [hunk 4]
```diff
@@ -2,18 +2,80 @@ import { SequenceMultiplicity, SequenceType, ValueType } from '../expressions/da
...
-	};
+	const steps = astHelper.getChildren(ast, 'stepExpr');
+	if (!steps) {
+		return { type: ValueType.ITEM, mult: SequenceMultiplicity.ZERO_OR_MORE };
+	}
+	let retType;
+	for (const step of steps) {
+		retType = annotateStep(step);
+	}
+	astHelper.insertAttribute(ast, 'type', retType);
```
Comment (FBroy) [inline src/typeInference/annotatePathExpr.ts:28]: Check if retType is not item()* when annotating the AST

**Review by FBroy — APPROVED** [review]

**Review by MTimmerije — APPROVED** [review]
Looks good to me.

**Pull request merged**